src/stridelayout.jl

```diff
@@ -274,36 +207,44 @@ An axis `i` of array `A` is dense if `stride(A, i) * Base.size(A, i) == stride(A
 """
 dense_dims(x) = dense_dims(typeof(x))
 dense_dims(::Type) = nothing
-_all_dense(::Val{N}) where {N} = DenseDims{ntuple(_ -> true, Val{N}())}()
+_all_dense(::Val{N}) where {N} = ntuple(_ -> True(), Val{N}())
+
 dense_dims(::Type{Array{T,N}}) where {T,N} = _all_dense(Val{N}())
-dense_dims(::Type{<:Tuple}) = DenseDims{(true,)}()
-function dense_dims(
-    ::Type{<:Union{Transpose{T,A},Adjoint{T,A}}},
-) where {T,A<:AbstractMatrix{T}}
-    dense = dense_dims(A)
-    isnothing(dense) ? nothing : dense[Val{(2, 1)}()]
+dense_dims(::Type{<:Tuple}) = (True(),)
+function dense_dims(::Type{T}) where {T<:MatAdjTrans}
+    dense = dense_dims(parent_type(T))
+    if dense === nothing
+        return nothing
+    else
+        return (last(dense), first(dense))
+    end
 end
-function dense_dims(
-    ::Type{<:PermutedDimsArray{T,N,I1,I2,A}},
-) where {T,N,I1,I2,A<:AbstractArray{T,N}}
+function dense_dims(::Type{<:PermutedDimsArray{T,N,I1,I2,A}}) where {T,N,I1,I2,A}
     dense = dense_dims(A)
-    isnothing(dense) ? nothing : dense[Val{I1}()]
+    if dense === nothing
+        return nothing
+    else
+        return permute(dense, Val(I1))
+    end
 end
 function dense_dims(::Type{S}) where {N,NP,T,A<:AbstractArray{T,NP},I,S<:SubArray{T,N,A,I}}
-    _dense_dims(S, dense_dims(A), stride_rank(A))
+    return _dense_dims(S, dense_dims(A), Val(stride_rank(A))) # TODO fix this
```

Comment:
What's wrong with this?

Reply:
nothing. It was an old note to myself that I never got rid of.